### ISSUE TITLE
jqPlot should not use elements' ID as targets

### ISSUE THREAD

**Issue by charlycoste** [jqPlot/jqPlot#6]
Developers may choose on their own, if they should pass an element ID, a DOMElement or a jQuery object to jqPlot:

For example:

    var chart1 = $('ul > li').get(2);
    var chart2 = $('#my-chart');
    $.jqplot(chart1, ...);
    $.jqplot(chart2, ...);

I'm working on a patch for it...

**Comment by dg-spark** [comment]
Im currently looking into this too, the code below will decode the target variable and return an jQuery object.

The line `if (l_object.nodeType)` replicates jQuery's handling of $(DOMElement)

I still working on investigating all instances of jqPlot referencing `$("#" + target)` so that it can be replaced by `$.jqplot.getTarget(target)`

Code:
```javascript
	$.jqplot.getTarget =  function(l_object) {
		if (typeof l_object == "string") return $("#" + l_object);
		if (l_object instanceof jQuery) return l_object;
		if (l_object.nodeType) return $(l_object);
		return $(l_object);
	}
```

Could probably be simplified to:??
```javascript
	$.jqplot.getTarget =  function(l_object) {
		if (typeof l_object == "string") return $("#" + l_object);
		return $(l_object);
	}
```

**Comment by charlycoste** [comment]
Actually, with your solution,`return $("#" + l_object)` will query the DOM each time you use `$.jqplot.getTarget`

I was more thinking about caching variables like BackboneJS does:

    // Creates the `this.el` and `this.$el` references for this view using the
    // given `el`. `el` can be a CSS selector or an HTML string, a jQuery
    // context or an element. Subclasses can override this to utilize an
    // alternative DOM manipulation API and are only required to set the
    // `this.el` property.
    _setElement: function(el) {
      this.$el = el instanceof Backbone.$ ? el : Backbone.$(el);
      this.el = this.$el[0];
    }

`Backbone.$` is just an alias to `jQuery` (or a compliant engine like zepto) so this is actually the same as : 

    _setElement: function(el) {
      this.$el = el instanceof $ ? el : $(el);
      this.el = this.$el[0];
    }

With this solution, you just need to use `this.el` if you want to do something on the DOM element, or use `this.$el` if you want to use the jQuery object.
$(el) will convert either a string selector or a DOM element to the good jquery object, so you don't need to use `typeof l_object` or `l_object.nodeType`

**Comment by dg-spark** [comment]
Yep thats quite right, however I only intened `getTarget` to be called once, during jqPlot.init. to set `this.target` check out `this.init = function(target, data, options)` in jqplot.core.js. It is caching the jquery object in the following:
```javascript
this.target = $('#'+target); 
```

The code you've provided definitely makes sense and could probably be used to assign to `this.target`

jqPlot also stores the targetId (string) which it appears is referenced by some plugins to locate the element. Either you would need to generate a unique `id` for  elements without an id or change the plugin interface.Hmmm.....

**Comment by charlycoste** [comment]
So you would use `this.target` and `this.$target` to let developers use the DOM or jQuery.
By the way, the problem of the unique `id` is the subject that gave me the most work for now.
I dont like the idea to add attributes on the DOM  when it's faster to deal with in memory objects than DOM tree (and also, easier to maintain).
But if it's required to avoid breaking compatibility with plugins, maybe the [unique-id module of jQuery-UI](https://github.com/jquery/jquery-ui/blob/master/ui/unique-id.js) can be a good example to start with.
